File: MapKitDemo/MapKitDemo/ViewController.m
//
//  ViewController.m
//  MapKitDemo
//
//  Created by Jecky on 14/11/13.
//  Copyright (c) 2014年 Jecky. All rights reserved.
//

#import "ViewController.h"
#import <MapKit/MapKit.h>
#import "RunData.h"
#import "MKUtil.h"
#import "RunAnnotationView.h"

#define kRunAnnotationStartPointTitle     @"起点"
#define kRunAnnotationEndPointTitle       @"终点"

@interface ViewController ()<MKMapViewDelegate,CLLocationManagerDelegate>{
    // the map view
    MKMapView* _mapView;
    
    // routes points
    NSMutableArray* _points;
    
    NSMutableArray  *_pointsPerKm;

    
    // the data representing the route points.
    MKPolyline* _routeLine;
    
    // the view we create for the line on the map
    MKPolylineView* _routeLineView;
    
    // the rect that bounds the loaded points
    MKMapRect _routeRect;
    
    // location manager
    CLLocationManager* _locationManager;
    
    // current location
    CLLocation* _currentLocation;
    
}

@property (nonatomic, retain) MKMapView  *mapView;
@property (nonatomic, retain) CLLocationManager *locationManager;
@property (nonatomic, retain) RunData *dataView;
@property (nonatomic, strong) UIButton       *locationButton;    //定位 中心

@property (nonatomic, strong) NSTimer          *runTimer;      //跑步计时器
@property (nonatomic, assign) CGFloat          totalMile;      //总路程
@property (nonatomic, assign) NSInteger        totalSecond;    //总时间
@property (nonatomic, assign) NSInteger        paceTime;       //记录配速的时间
@property (nonatomic, assign) NSInteger        paceMile;       //记录配速的距离

@property (nonatomic, assign) BOOL        isRunning;
@property (nonatomic, assign) BOOL        startPointDrawed;     //开始点是否被绘制

@property (nonatomic, assign) MKMapPoint  northEastPoint;
@property (nonatomic, assign) MKMapPoint  southWestPoint;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    self.navigationItem.title = @"MapKitDemo";
    if ([self respondsToSelector:@selector(automaticallyAdjustsScrollViewInsets)]) {
        self.automaticallyAdjustsScrollViewInsets = NO;
    }
    
    [self dataView];
    [self mapView];
    [self locationButton];
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"开始" style:UIBarButtonItemStylePlain target:self action:@selector(start)];
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"暂停" style:UIBarButtonItemStylePlain target:self action:@selector(pauseRun)];

    _locationManager = [[CLLocationManager alloc] init];
    _locationManager.pausesLocationUpdatesAutomatically = NO;
    if ([_locationManager respondsToSelector:@selector(requestWhenInUseAuthorization)]){
        [_locationManager requestWhenInUseAuthorization];
    }
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(stopUpdatingLocation)
                                                 name:UIApplicationDidEnterBackgroundNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(startUpdatingLocation)
                                                 name:UIApplicationWillEnterForegroundNotification
                                               object:nil];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    
    //提示用户打开定位服务
    if (![CLLocationManager locationServicesEnabled]|| [CLLocationManager authorizationStatus] == kCLAuthorizationStatusDenied){
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"系统提示"
                                                        message:@"位置服务不可用，请先进入设置-隐私中开启定位服务"
                                                       delegate:self
                                              cancelButtonTitle:@"我知道了"
                                              otherButtonTitles:nil];
        [alert show];
    }else{
        [self startUpdatingLocation];
    }
}

- (void)startUpdatingLocation
{
    [_locationManager startUpdatingLocation];
}

- (void)stopUpdatingLocation
{
    if (_isRunning) {
        return;
    }
    
    [_locationManager stopUpdatingLocation];
}

- (void)showUserLocationCenter
{
    if (self.mapView.userLocation.coordinate.latitude == 0.0
        &&self.mapView.userLocation.coordinate.longitude == 0.0) {
        NSLog(@"========没有定位点========");
        return;
    }
    
    [self.mapView setCenterCoordinate:self.mapView.userLocation.coordinate animated:YES];
}

- (void)start{
    NSLog(@"=========开始==========");
    if (_isRunning) {
        return;
    }
    
    _isRunning = YES;
    _points = [NSMutableArray arrayWithCapacity:0];
    _pointsPerKm = [NSMutableArray arrayWithCapacity:0];
    self.totalSecond = 0;
    self.totalMile = 0;
    
    if ([_runTimer isValid]) {
        [_runTimer invalidate];
        _runTimer = nil;
    }
    _runTimer = [NSTimer scheduledTimerWithTimeInterval:1
                                                 target:self
                                               selector:@selector(updateRunStatusView)
                                               userInfo:nil
                                                repeats:YES];
    
    [_locationManager startUpdatingLocation];
}


- (void)pauseRun
{
    [_runTimer invalidate];
    _runTimer = nil;
    _isRunning = NO;
    
    
    if ([_points count]>1) {
        CLLocation *location = [_points lastObject];
        [self addPointAnnotation:CLLocationCoordinate2DMake(location.coordinate.latitude, location.coordinate.longitude) title:kRunAnnotationEndPointTitle];
        
    }
}

- (void)updateRunStatusView
{
    [_locationManager startUpdatingLocation];

    if(_paceMile != (NSInteger)floor(_totalMile/100.0)){
        _paceMile = (NSInteger)floor(_totalMile/100.0);
        
        CLLocation  *location = [_points lastObject];
        if (location) {
            [_pointsPerKm addObject:location];
            NSString  *locat= [NSString stringWithFormat:@"%lu",(unsigned long)[_pointsPerKm count]];
            [self addPointAnnotation:location.coordinate title:locat];
        }

        _paceTime = _totalSecond;
    }

    self.totalSecond++;
    
    [self.dataView updateRunData:_totalSecond mile:_totalMile];
}

- (void)configureRoutes
{
    // define minimum, maximum points
    _northEastPoint = MKMapPointMake(0.f, 0.f);
    _southWestPoint = MKMapPointMake(0.f, 0.f);
    
    // create a c array of points.
    MKMapPoint* pointArray = malloc(sizeof(CLLocationCoordinate2D) * _points.count);
    
    for(int idx = 0; idx < _points.count; idx++)
    {
        CLLocation *location = [_points objectAtIndex:idx];
        CLLocationDegrees latitude  = location.coordinate.latitude;
        CLLocationDegrees longitude = location.coordinate.longitude;
        
        // create our coordinate and add it to the correct spot in the array
        CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake(latitude, longitude);
        MKMapPoint point = MKMapPointForCoordinate(coordinate);
        
        // if it is the first point, just use them, since we have nothing to compare to yet.
        if (idx == 0) {
            _northEastPoint = point;
            _southWestPoint = point;
        } else {
            if (point.x > _northEastPoint.x)
                _northEastPoint.x = point.x;
            if(point.y > _northEastPoint.y)
                _northEastPoint.y = point.y;
            if (point.x < _southWestPoint.x)
                _southWestPoint.x = point.x;
            if (point.y < _southWestPoint.y)
                _southWestPoint.y = point.y;
        }
        
        pointArray[idx] = point;
    }
    
    if (_routeLine) {
        [self.mapView removeOverlay:_routeLine];
    }
    
    _routeLine = [MKPolyline polylineWithPoints:pointArray count:_points.count];
    
    // add the overlay to the map
    if (nil != _routeLine) {
        [self.mapView addOverlay:_routeLine];
    }
    
    // clear the memory allocated earlier for the points
    free(pointArray);
    
    /*
     double width = northEastPoint.x - southWestPoint.x;
     double height = northEastPoint.y - southWestPoint.y;
     
     _routeRect = MKMapRectMake(southWestPoint.x, southWestPoint.y, width, height);    	
     
     // zoom in on the route. 
     [self.mapView setVisibleMapRect:_routeRect];
     */
}

- (void)showInMapView:(CLPlacemark *) placemark
{
    CLLocationCoordinate2D coordinate = placemark.location.coordinate;
    // 添加MapView
    MKCoordinateSpan span = MKCoordinateSpanMake(0.01, 0.01);			// 跨度（比例）
    MKCoordinateRegion region = MKCoordinateRegionMake(coordinate, span);       // 范围、区域
    [self.mapView setRegion:region];
}

-(void)addPointAnnotation:(CLLocationCoordinate2D)coordinate title:(NSString *)titile
{
    MKPointAnnotation *annotation = [[MKPointAnnotation alloc] init];
    annotation.coordinate = coordinate;
    annotation.title    = titile;
    [self.mapView addAnnotation:annotation];
}

#pragma mark - MKMapViewDelegate
- (void)mapView:(MKMapView *)mapView didUpdateUserLocation:(MKUserLocation *)userLocation{

    if (!_isRunning) {
        return;
    }
    
    CLLocation *location = [[CLLocation alloc] initWithLatitude:userLocation.coordinate.latitude
                                                      longitude:userLocation.coordinate.longitude];
    // check the zero point
    if  (userLocation.coordinate.latitude == 0.0f ||
         userLocation.coordinate.longitude == 0.0f)
        return;
    
    //坐标点为同一点
    if (location.coordinate.latitude == _currentLocation.coordinate.latitude
        &&location.coordinate.longitude == _currentLocation.coordinate.longitude){
        return;
    }
    
    if(userLocation.location.horizontalAccuracy <70 && userLocation.location.verticalAccuracy<70){
        if(_totalSecond<3)//前三秒不取点
            return;
        
        if (_points.count > 0) {
            CLLocationDistance distance = [location distanceFromLocation:_currentLocation];
            if(distance>0){
                _totalMile = _totalMile + distance;
            }
        }
        //两秒取一次用户位置
        if(_totalSecond%5 == 0){
            if(_points.count<_totalSecond){ //保证每秒获取一个经纬度信息点
                [_points addObject:location];
                [self configureRoutes];
            }
        }

        if ([_points count]==1&&!_startPointDrawed) {
            _startPointDrawed = YES;
            [self addPointAnnotation:location.coordinate title:kRunAnnotationStartPointTitle];
        }
        _currentLocation = location;
        
    }
    
}

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id )annotation
{
    if ([annotation isKindOfClass:[MKPointAnnotation class]]){

        static NSString *CustomAnnotationReuseIndetifier = @"CustomAnnotationReuseIndetifier";
        RunAnnotationView *annotationView = (RunAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:CustomAnnotationReuseIndetifier];
        if (annotationView == nil) {
            annotationView = [[RunAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:CustomAnnotationReuseIndetifier];
        }
        
        if ([[annotation title] isEqualToString:kRunAnnotationStartPointTitle]){ /* 起点. */
            annotationView.annotationImageView.image = [UIImage imageNamed:@"bg_run_start_point"];
        }else if([[annotation title] isEqualToString:kRunAnnotationEndPointTitle]){ /* 终点. */
            annotationView.annotationImageView.image = [UIImage imageNamed:@"bg_run_end_point"];
        }else{
            annotationView.annotationImageView.image = [UIImage imageNamed:@"bg_run_inter_point"];
            annotationView.nameLabel.text    = [annotation title];
        }
        
        return annotationView;
    }
    return nil;
    

}

- (MKOverlayView *)mapView:(MKMapView *)mapView viewForOverlay:(id <MKOverlay>)overlay
{
    NSLog(@"%@ ----- %@", self, NSStringFromSelector(_cmd));
    MKOverlayView* overlayView = nil;
    if(overlay == _routeLine)
    {
        //if we have not yet created an overlay view for this overlay, create it now.
        if (_routeLineView) {
            [_routeLineView removeFromSuperview];
        }
        
        _routeLineView = [[MKPolylineView alloc] initWithPolyline:_routeLine];
        _routeLineView.fillColor = [UIColor orangeColor];
        _routeLineView.strokeColor = [UIColor orangeColor];
        _routeLineView.lineWidth = 4;
        
        overlayView = _routeLineView;
    }
    
    return overlayView;
}

#pragma mark - getter
- (MKMapView *)mapView{
    if (!_mapView) {
        _mapView = [[MKMapView alloc] initWithFrame:CGRectMake(0, CGRectGetMaxY(self.dataView.frame)-10, self.view.bounds.size.width, self.view.bounds.size.height-CGRectGetMaxY(self.dataView.frame)+10)];
        _mapView.delegate = self;
        _mapView.showsUserLocation = YES;
        _mapView.userInteractionEnabled = YES;
        _mapView.zoomEnabled = YES;
        _mapView.rotateEnabled = NO;
        _mapView.showsBuildings = YES;
        //setup default display map region
        _mapView.region = MKCoordinateRegionMake(CLLocationCoordinate2DMake(31.27189408, 121.47583771), MKCoordinateSpanMake(0.001, 0.001));
        _mapView.userTrackingMode = MKUserTrackingModeFollowWithHeading;
        [self.view insertSubview:_mapView belowSubview:self.dataView];
    }
    return _mapView;
}

- (RunData *)dataView{
    if (!_dataView) {
        _dataView = [[RunData alloc] initWithFrame:CGRectMake(0, 64, self.view.bounds.size.width, 157)];
        [self.view addSubview:_dataView];

    }
    return _dataView;
}

- (UIButton *)locationButton{
    if (!_locationButton) {
        UIImage  *img = [UIImage imageNamed:@"btn_fix_nor"];
        _locationButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _locationButton.frame = CGRectMake(0, self.view.bounds.size.height-140, img.size.width, img.size.height);
        [_locationButton setBackgroundImage:img forState:UIControlStateNormal];
        [_locationButton setBackgroundImage:[UIImage imageNamed:@"btn_fix_press"] forState:UIControlStateSelected];
        _locationButton.exclusiveTouch = YES;
        [_locationButton addTarget:self action:@selector(showUserLocationCenter) forControlEvents:UIControlEventTouchUpInside];
        [self.view addSubview:_locationButton];
    }
    return _locationButton;
}

@end

File: MapKitSwift/MapKitSwift/ViewController.swift
//
//  ViewController.swift
//  MapKitSwift
//
//  Created by Jecky on 14/11/13.
//  Copyright (c) 2014年 Jecky. All rights reserved.
//

import UIKit
import MapKit

class ViewController: UIViewController, MKMapViewDelegate {
    
    var mapView:MKMapView?
    var dataView:RunDataView?
    var locationButton:UIButton?
    
    var startPointDrawed:Bool!
    var isRunning:Bool!
    
    var runTimer:NSTimer?
    var totalMile:Double!       = 0.0      //跑步距离
    var totalSecond:NSInteger?   = 0      //跑步时间
    var paceTime:NSInteger?      = 0      //记录配速的时间
    var paceMile:NSInteger?      = 0      //记录配速的距离
    
    var currentLocation:CLLocation?
    var locationManager:CLLocationManager?
    var points:NSMutableArray?
    var pointsPerKm:NSMutableArray!

    // the data representing the route points.
    var routeLine:MKPolyline?
    
    // the view we create for the line on the map
    var routeLineView:MKPolylineRenderer?
    
    var northEastPoint:MKMapPoint?
    var southWestPoint:MKMapPoint?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.navigationItem.title = "MapKitSwift"
        
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "开始", style: UIBarButtonItemStyle.Plain, target: self, action: "start")
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "暂停", style: UIBarButtonItemStyle.Plain, target: self, action: "pauseRun")
        
        locationManager = CLLocationManager()
        if locationManager!.respondsToSelector("requestWhenInUseAuthorization") {
            locationManager!.requestWhenInUseAuthorization()
        }
        
        isRunning = false
        startPointDrawed = false
        
        initUI()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "stopUpdatingLocation", name: UIApplicationDidEnterBackgroundNotification, object: nil)
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "startUpdatingLocation", name: UIApplicationWillEnterForegroundNotification, object: nil)
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        if !CLLocationManager.locationServicesEnabled() || CLLocationManager.authorizationStatus() == CLAuthorizationStatus.Denied {
            
            var alert = UIAlertView()
            alert.title = "系统提示"
            alert.message = "位置服务不可用，请先进入设置-隐私中开启定位服务"
            alert.addButtonWithTitle("我知道了")
            alert.show()
        }else{
            locationManager!.startUpdatingLocation()
        }
        

    }
    
    func startUpdatingLocation() {
        locationManager!.startUpdatingLocation()
    }
    
    func stopUpdatingLocation() {
        if isRunning! {
            return
        }
        
        locationManager!.stopUpdatingLocation()
    }
    
    func showUserLocationCenter() {
        if mapView!.userLocation.coordinate.latitude == 0.0&&mapView!.userLocation.coordinate.longitude == 0.0 {
            NSLog("=======NO Location=========")
            return
        }
        self.mapView!.setCenterCoordinate(self.mapView!.userLocation.coordinate, animated: true)
    }
    
    // TODO:
    //MARK: -----Run------
    func start() {
        NSLog("======== Start ============")
        if (isRunning!) {
            return
        }
        
        isRunning = true
        points = NSMutableArray(capacity: 0)
        pointsPerKm = NSMutableArray(capacity: 0)
        
        self.totalSecond = 0
        self.totalMile = 0.0
        
        setRunTimer()
        
        locationManager?.startUpdatingLocation()
    }
    
    func pauseRun() {
        NSLog("======== pauseRun ============")
        
        isRunning = false
        runTimer?.invalidate()
        runTimer = nil
        
        if (points!.count>1){
            var location = points!.lastObject as CLLocation
            addPointAnnotation(location.coordinate, title: "终点")
        }
    }
    
    func setRunTimer(){
        if (runTimer != nil) {
            runTimer!.invalidate()
            runTimer = nil
        }
        runTimer = NSTimer.scheduledTimerWithTimeInterval(1.0, target: self, selector: "updateRunStatusView", userInfo: nil, repeats: true)
    }
    
    func updateRunStatusView() {
        locationManager!.startUpdatingLocation()

        let integerPi = Int(totalMile!/1000.0)
        
        if paceMile != integerPi {
            paceMile = integerPi
            
            var location = points?.lastObject as CLLocation
            if (points?.count > 0) {
                pointsPerKm.addObject(location)
                addPointAnnotation(location.coordinate, title: "\(pointsPerKm.count)")
            }

            paceTime = totalSecond
        }
        
        totalSecond!++
        
        dataView?.updateRunData(totalSecond!, totalMile: totalMile!)
    }
    
    func configureRoutes() {
        // define minimum, maximum points
        northEastPoint = MKMapPointMake(0.0, 0.0)
        southWestPoint = MKMapPointMake(0.0, 0.0)
        
        var pointArray: UnsafeMutablePointer<MKMapPoint> = UnsafeMutablePointer.alloc(points!.count)
        
        for (var idx = 0; idx < points!.count; idx++){
            var location:CLLocation = points!.objectAtIndex(idx) as CLLocation
            // create our coordinate and add it to the correct spot in the array
            var point = MKMapPointForCoordinate(CLLocationCoordinate2DMake(location.coordinate.latitude, location.coordinate.longitude)) as MKMapPoint
            
            // if it is the first point, just use them, since we have nothing to compare to yet.
            if (idx == 0) {
                northEastPoint = point;
                southWestPoint = point;
            } else {
                if (point.x > northEastPoint!.x){
                    northEastPoint!.x = point.x;
                }
                if(point.y > northEastPoint!.y){
                    northEastPoint!.y = point.y;
                }
                if (point.x < southWestPoint!.x){
                    southWestPoint!.x = point.x;
                }
                if (point.y < southWestPoint!.y){
                    southWestPoint!.y = point.y;
                }
            }
            
            pointArray[idx] = point
        }

        if (routeLine != nil) {
            self.mapView?.removeOverlay(routeLine)
        }
        
        routeLine = MKPolyline(points: pointArray, count: points!.count)

        if (routeLine != nil) {
            self.mapView?.addOverlay(routeLine)
        }
        free(pointArray);
        
    }
    
    func addPointAnnotation(coordinate:CLLocationCoordinate2D, title:NSString){
        var annotation:MKPointAnnotation = MKPointAnnotation()
        annotation.coordinate = coordinate
        annotation.title = title
        self.mapView?.addAnnotation(annotation)
    }
    
    // TODO:
    //MARK: -----MKMapViewDelegate------
    func mapView(mapView: MKMapView!, didUpdateUserLocation userLocation: MKUserLocation!){
        
        if (!isRunning) {
            return
        }
        
        if(userLocation.coordinate.latitude == 0.0 || userLocation.coordinate.longitude == 0.0){
            return
        }
        
        var location:CLLocation! = CLLocation(latitude: userLocation.coordinate.latitude, longitude: userLocation.coordinate.longitude)
        
        if userLocation.location.horizontalAccuracy < 70 && userLocation.location.verticalAccuracy < 70{
            if totalSecond! < 3 {
                return
            }
            
            if points!.count > 0 {
                var distance:CLLocationDistance = location.distanceFromLocation(self.currentLocation)
                if distance>0 {
                    totalMile! = totalMile! + distance
                }
            }
            
            if totalSecond!%2 == 0 {
                if points!.count < totalSecond! {
                    points?.addObject(location)
                    configureRoutes()
                }
            }
            
            if ((points!.count==1)&&(!startPointDrawed!)){
                startPointDrawed = true
                addPointAnnotation(location.coordinate, title: "起点")
            }
            self.currentLocation = location;
        }
    }
    
    func mapView(mapView: MKMapView!, viewForAnnotation annotation: MKAnnotation!) -> MKAnnotationView! {
        
        if annotation.isKindOfClass(MKPointAnnotation) {
            let CustomAnnotationReuseIndetifier = "CustomAnnotationReuseIndetifier"

            var annotationView = mapView.dequeueReusableAnnotationViewWithIdentifier(CustomAnnotationReuseIndetifier) as? CustomAnnotationView
            if annotationView == nil {
                annotationView = CustomAnnotationView(annotation: annotation, reuseIdentifier: CustomAnnotationReuseIndetifier)
                annotationView?.bounds = CGRectMake(0.0, 0.0, 26.0, 56.0)
                annotationView?.backgroundColor = UIColor.clearColor()
            }
            
            if (annotation.title == "起点") {
                annotationView!.annotationImageView.image = UIImage(named: "bg_run_start_point")
            }else if (annotation.title == "终点") {
                annotationView!.annotationImageView.image = UIImage(named: "bg_run_end_point")
            }else{
                annotationView!.annotationImageView.image = UIImage(named: "bg_run_inter_point")
                annotationView!.nameLabel.text = annotation.title
            }
            return annotationView;
            
        }

        return nil
    }

    func mapView(mapView: MKMapView!, rendererForOverlay overlay: MKOverlay!) -> MKOverlayRenderer!{
        if overlay is MKPolyline {
            var renderer:MKPolylineRenderer = MKPolylineRenderer(polyline: routeLine!)
            renderer.fillColor = UIColor.orangeColor()
            renderer.strokeColor = UIColor.orangeColor()
            renderer.lineWidth = 4
            return renderer
        }
        
        return nil
    }
    
    // TODO:
    //MARK: -----Init UI------
    func initUI() {
        dataView = RunDataView(frame: CGRectMake(0, 64, self.view.bounds.size.width, 157))
        self.view.addSubview(dataView!)
        
        mapView = MKMapView(frame: CGRectMake(0, CGRectGetMaxY(dataView!.frame)-10, self.view.bounds.size.width, self.view.bounds.size.height-CGRectGetMaxY(dataView!.frame)+10))
        mapView?.delegate = self
        mapView?.showsUserLocation = true
        mapView?.userInteractionEnabled = true
        mapView?.zoomEnabled = true
        mapView?.rotateEnabled = false
        mapView?.showsBuildings = true
        mapView?.region = MKCoordinateRegionMake(CLLocationCoordinate2DMake(31.27189408, 121.47583771), MKCoordinateSpanMake(0.001, 0.001))
        mapView?.userTrackingMode = MKUserTrackingMode.FollowWithHeading
        self.view.insertSubview(mapView!, belowSubview: dataView!)
        
        var img:UIImage! = UIImage(named: "btn_fix_nor")
        locationButton = UIButton.buttonWithType(UIButtonType.Custom) as? UIButton
        locationButton?.frame = CGRectMake(0, self.view.bounds.size.height-140, img.size.width, img.size.height)
        locationButton?.setBackgroundImage(img, forState: UIControlState.Normal)
        locationButton?.setBackgroundImage(UIImage(named: "btn_fix_press"), forState: UIControlState.Selected)
        locationButton?.exclusiveTouch = true
        locationButton?.addTarget(self, action: "showUserLocationCenter", forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(locationButton!)

    }

}

